Doxygen/Docs/man/man3/Engine.3
.TH "Engine" 3 "Fri Mar 29 2013" "Version 31337" "HyperGrafx" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Engine \- 
.PP
The \fBEngine\fP class is a singleton-style class which helps keep track of instances of important objects (for \fBCameras\fP, Objects, etc) as well as some settings and variables that would otherwise clog up global namespace\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Engine\&.hpp>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCameras\fP * \fBcams\fP (void)"
.br
.RI "\fIRetrieves a pointer to the \fBCamera\fP List\&. \fP"
.ti -1c
.RI "\fBScene\fP * \fBrootScene\fP (void)"
.br
.RI "\fIRetrieves a pointer to the Root, top-level \fBScene\fP graph\&. \fP"
.ti -1c
.RI "\fBScreen\fP * \fBmainScreen\fP (void)"
.br
.RI "\fIRetrieves a pointer to the core '\fBScreen\fP' object\&. \fP"
.ti -1c
.RI "bool \fBopt\fP (const std::string &Option)"
.br
.RI "\fIopt retrieves the current setting of an option in the \fBEngine\fP\&. \fP"
.ti -1c
.RI "void \fBopt\fP (const std::string &Option, bool setting)"
.br
.RI "\fIopt, with a second parameter, sets an \fBEngine\fP option\&. \fP"
.ti -1c
.RI "bool \fBset\fP (const std::string &Option)"
.br
.RI "\fIset checks to see if an option has been explicitly set to either True/False\&. \fP"
.ti -1c
.RI "bool \fBflip\fP (const std::string &Option)"
.br
.RI "\fIflip changes a setting from its current value to its negated form\&. \fP"
.ti -1c
.RI "\fB~Engine\fP (void)"
.br
.RI "\fIDefault, non-virtual destructor\&. \fP"
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBEngine\fP * \fBinstance\fP (void)"
.br
.RI "\fIinstance returns, or creates and then returns, a pointer to the \fBEngine\fP object\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "\fBEngine\fP (void)"
.br
.RI "\fIDefault constructor\&. \fP"
.ti -1c
.RI "\fBEngine\fP (const \fBEngine\fP &copy)"
.br
.RI "\fICopy constructor\&. \fP"
.ti -1c
.RI "\fBEngine\fP & \fBoperator=\fP (\fBEngine\fP &assign)"
.br
.RI "\fIAssignment operator\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBScene\fP \fB_scene\fP"
.br
.RI "\fIThe root \fBScene\fP graph for the \fBEngine\fP\&. \fP"
.ti -1c
.RI "\fBScreen\fP \fB_screen\fP"
.br
.RI "\fIThe core 'Screen' object for the \fBEngine\fP\&. \fP"
.ti -1c
.RI "\fBSettingsMap\fP \fB_engineSettings\fP"
.br
.RI "\fI_engineSettings is a std::map that contains a series of <std::string, bool> pairs that represent our \fBEngine\fP options\&. \fP"
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static \fBEngine\fP * \fB_engineSingleton\fP = NULL"
.br
.RI "\fIstatic, stateful variable that is our singleton pointer\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
The \fBEngine\fP class is a singleton-style class which helps keep track of instances of important objects (for \fBCameras\fP, Objects, etc) as well as some settings and variables that would otherwise clog up global namespace\&. 

\fBAuthor:\fP
.RS 4
John Huston, jhuston@cs.uml.edu 
.RE
.PP
\fBDate:\fP
.RS 4
2013-03-13 
.RE
.PP

.PP
Definition at line 32 of file Engine\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Engine::Engine (void)\fC [private]\fP"

.PP
Default constructor\&. Cannot be called, this is a singleton class\&. 
.PP
Definition at line 40 of file Engine\&.cpp\&.
.SS "Engine::Engine (const \fBEngine\fP &copy)\fC [private]\fP"

.PP
Copy constructor\&. Cannot be called\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcopy\fP Not used\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Will always throw an exception\&. 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIWill\fP always throw std::logic_error\&. 
.RE
.PP

.PP
Definition at line 57 of file Engine\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBCameras\fP * Engine::cams (void)"

.PP
Retrieves a pointer to the \fBCamera\fP List\&. \fBReturns:\fP
.RS 4
A pointer to the \fBCamera\fP List\&. 
.RE
.PP

.PP
Definition at line 76 of file Engine\&.cpp\&.
.SS "bool Engine::flip (const std::string &Option)"

.PP
flip changes a setting from its current value to its negated form\&. \fBParameters:\fP
.RS 4
\fIOption\fP The option to toggle\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
The new, current value of the option\&. 
.RE
.PP

.PP
Definition at line 139 of file Engine\&.cpp\&.
.SS "\fBEngine\fP * Engine::instance (void)\fC [static]\fP"

.PP
instance returns, or creates and then returns, a pointer to the \fBEngine\fP object\&. All hail the singleton!
.PP
\fBReturns:\fP
.RS 4
A pointer to the \fBEngine\fP object\&. 
.RE
.PP

.PP
Definition at line 29 of file Engine\&.cpp\&.
.SS "\fBScreen\fP * Engine::mainScreen (void)"

.PP
Retrieves a pointer to the core '\fBScreen\fP' object\&. \fBReturns:\fP
.RS 4
A pointer to the core '\fBScreen\fP' object\&. 
.RE
.PP

.PP
Definition at line 96 of file Engine\&.cpp\&.
.SS "\fBEngine\fP & Engine::operator= (\fBEngine\fP &assign)\fC [private]\fP"

.PP
Assignment operator\&. Cannot be used\&. This is a singleton class\&. 
.PP
\fBParameters:\fP
.RS 4
\fIassign\fP Not used\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Will always throw an exception\&. 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIWill\fP always throw std::logic_error\&. 
.RE
.PP

.PP
Definition at line 67 of file Engine\&.cpp\&.
.SS "bool Engine::opt (const std::string &Option)"

.PP
opt retrieves the current setting of an option in the \fBEngine\fP\&. \fBParameters:\fP
.RS 4
\fIOption\fP The name of the option to access\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
A bool: the current value of the setting\&. 
.RE
.PP

.PP
Definition at line 107 of file Engine\&.cpp\&.
.SS "void Engine::opt (const std::string &Option, boolsetting)"

.PP
opt, with a second parameter, sets an \fBEngine\fP option\&. \fBParameters:\fP
.RS 4
\fIOption\fP The name of the option to set\&. 
.br
\fIsetting\fP The value to give the option\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
void\&. 
.RE
.PP

.PP
Definition at line 117 of file Engine\&.cpp\&.
.SS "\fBScene\fP * Engine::rootScene (void)"

.PP
Retrieves a pointer to the Root, top-level \fBScene\fP graph\&. \fBReturns:\fP
.RS 4
A pointer to the Root-level \fBScene\fP graph\&. 
.RE
.PP

.PP
Definition at line 86 of file Engine\&.cpp\&.
.SS "bool Engine::set (const std::string &Option)"

.PP
set checks to see if an option has been explicitly set to either True/False\&. \fBParameters:\fP
.RS 4
\fIOption\fP The option to check the existence of 
.RE
.PP
\fBReturns:\fP
.RS 4
A boolean: True if the option has been set, False otherwise\&. 
.RE
.PP

.PP
Definition at line 127 of file Engine\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for HyperGrafx from the source code\&.


src/Classes/Engine.cpp
/**
 @file Engine.cpp
 @author John Huston
 @authors https://github.com/UMLComputerGraphics
 @date 2013-03-13
 @brief Implementation for the Engine class.
 **/
#include <stdexcept>
#include <string>
#include <map>

#include "Engine.hpp"
#include "Cameras.hpp"
#include "Scene.hpp"
#include "Screen.hpp"
#include "glut_callbacks.h"

/**
 * static, stateful variable that is our singleton pointer.
 */
Engine *Engine::_engineSingleton = NULL;

/**
 * instance returns, or creates and then returns,
 * a pointer to the Engine object.
 * All hail the singleton!
 *
 * @return A pointer to the Engine object.
 */
Engine *Engine::instance( void ) {
  
  if ( _engineSingleton == NULL ) _engineSingleton = new Engine();
  
  return _engineSingleton;
  
}

/**
 * Default constructor. Cannot be called, this is a singleton class.
 */
Engine::Engine( void ) {
  _idleFunc = NULL;
}

/**
 * Default, non-virtual destructor.
 */
Engine::~Engine( void ) {
  // Nihil.
}

/**
 * Copy constructor. Cannot be called.
 * @param copy Not used.
 * @return Will always throw an exception.
 * @throw Will always throw std::logic_error.
 */
Engine::Engine( const Engine &copy ) {
  throw std::logic_error( "Engine is a Singleton. You can't make more!" );
}

/**
 * Assignment operator. Cannot be used. This is a singleton class.
 * @param assign Not used.
 * @return Will always throw an exception.
 * @throw Will always throw std::logic_error.
 */
Engine &Engine::operator=( Engine &assign ) {
  throw std::logic_error(
      "Cannot assign a Engine to another; You can seriously only have one!" );
}

/**
 * Retrieves a pointer to the Camera List.
 * @return A pointer to the Camera List.
 */
Cameras *Engine::cams( void ) {
  
  return &(_screen._camList);
  
}

/**
 * Retrieves a pointer to the Root, top-level Scene graph.
 * @return A pointer to the Root-level Scene graph.
 */
Scene *Engine::rootScene( void ) {
  
  return &_scene;
  
}

/**
 * Retrieves a pointer to the core 'Screen' object.
 * @return A pointer to the core 'Screen' object.
 */
Screen *Engine::mainScreen( void ) {
  
  return &_screen;
  
}

/**
 * Retrieves a pointer to the Texture Management object.
 * @return A pointer to the Texture Management object.
 */
TextureManagement *Engine::texMan( void ){
  return &_texMan;
}

/**
 * opt retrieves the current setting of an option in the Engine.
 * @param Option The name of the option to access.
 * @return A bool: the current value of the setting.
 */
bool Engine::opt( const std::string &Option ) {
  return _engineSettings[Option];
}

/**
 * opt, with a second parameter, sets an Engine option.
 * @param Option The name of the option to set.
 * @param setting The value to give the option.
 * @return void.
 */
void Engine::opt( const std::string &Option, bool setting ) {
  _engineSettings[Option] = setting;
}

/**
 * set checks to see if an option has been explicitly set
 * to either True/False.
 * @param Option The option to check the existence of
 * @return A boolean: True if the option has been set, False otherwise.
 */
bool Engine::set( const std::string &Option ) {
  SettingsMap::iterator it;
  it = _engineSettings.find( Option );
  if ( it == _engineSettings.end() ) return false;
  return true;
}

/**
 * flip changes a setting from its current value to its negated form.
 * @param Option The option to toggle.
 * @return The new, current value of the option.
 */
bool Engine::flip( const std::string &Option ) {
  return _engineSettings[Option] = (!_engineSettings[Option]);
}

/**
 * Initialize GLEW, GLUT and our Engine.
 */
void Engine::init( int *argc, char *argv[], const char *title ) {

  // OS X suppresses events after mouse warp.  This resets the suppression
  // interval to 0 so that events will not be suppressed. This also found
  // at http://stackoverflow.com/questions/728049/
  // glutpassivemotionfunc-and-glutwarpmousepointer
#ifdef __APPLE__
  CGSetLocalEventsSuppressionInterval( 0.0 );
#endif
  Util::InitRelativePaths(*argc, argv);
  glutInit( argc, argv );
  glutInitDisplayMode( GLUT_RGBA | GLUT_DOUBLE | GLUT_DEPTH );
  glutInitWindowSize( 0, 0 );
  glutCreateWindow( title );
  glutFullScreen();
  glutSetCursor( GLUT_CURSOR_NONE );

  GLEW_INIT();

  glutCreateWindow( "Linear Interpolation Morphing Demo" );
  glutFullScreen();
  glutSetCursor( GLUT_CURSOR_NONE );

  /* Register our Callbacks */
  glutDisplayFunc( Engine::displayScreen );
  glutKeyboardFunc( engineKeyboard );
  glutKeyboardUpFunc( engineKeylift );
  glutSpecialFunc( engineSpecialKeyboard );
  glutMouseFunc( engineMouse );
  glutMotionFunc( engineMouseMotion );
  glutPassiveMotionFunc( EngineMousePassive );
  glutIdleFunc( Engine::idle );
  glutReshapeFunc( engineResize );

}


void Engine::registerIdle( void (idleFunc)( void ) ) {
  _idleFunc = idleFunc;
}


/**
 * What should the engine be doing every idle()?
 */
void Engine::idle( void ) {

  static Cameras *camList = Engine::instance()->cams();

  // Compute the time since last idle().
  tick.tock();

  // Move all camera(s).
  camList->idleMotion();

  Engine::instance()->callIdle();

  glutPostRedisplay();

}

void Engine::callIdle( void ) {
  if (_idleFunc) (*_idleFunc)();
}


/**
 * What should we be doing to display()?
 * Display/render the entire screen.
 */
void Engine::displayScreen( void ) {

  static Cameras *camList = Engine::instance()->cams();

  glClear( GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT );

  // Tell camList to draw using our 'draw' rendering function.
  camList->view( Engine::displayViewport );

  glutSwapBuffers();

}

/**
 * Display/re-render a viewport.
 */
void Engine::displayViewport( void ) {

  static Scene *theScene = Engine::instance()->rootScene();
  static Cameras *camList = Engine::instance()->cams();

  theScene->draw();
  camList->draw();

}
